File: rustversion/src/board.rs
use crate::cell::{self, Cell}; 
use rand::Rng;



#[derive(Debug)]
pub struct Board {
    rows: i32,
    cols: i32,
    grid: Vec<Vec<Cell>>,
}

pub fn generate_board(instantiated_board: Board) -> Board {
    let mut new_board = instantiated_board;
    for row in 0..new_board.rows {
        for col in 0..new_board.cols {
            let chance_number = rand::thread_rng().gen_range(0 as i32, 3 as i32);
            
            let mut new_cell: Cell = cell::get_new_cell(); 

            // 33 chance to spawn
            if chance_number == 1 {
                new_cell.set_alive();
            } else {
                new_cell.set_dead();
            }

            new_board.grid[row as usize][col as usize] = new_cell
        }
    }

    return new_board 
}

pub fn get_new_board(rows: i32, cols: i32) -> Board {
    let fake_cell: Cell = cell::get_new_cell();
    let instantiated_board= Board {
        rows: rows,
        cols: cols,
        grid: vec![vec![fake_cell; cols as usize]; rows as usize],
    };

    let generated_board = generate_board(instantiated_board);
    
    return generated_board;
}

impl Board {
    pub fn show(&self) {
        for row in 0..self.rows {
            for col in 0..self.cols {
               self.grid[row as usize][col as usize].show();
            }
            println!();
        }
    }
    
    pub fn update(&mut self) {
        for row in 0..self.rows {
            for col in 0..self.cols {
                let mut cell = self.grid[row as usize][col as usize];
                let (alive_neighbors, mut _neighbors) = self.get_cell_neighbor_info(cell);
                
                // conways rules!
                if cell.is_alive() {
                   if alive_neighbors.len() < 2 {
                    cell.set_dead();
                   }  else if alive_neighbors.len() == 2 || alive_neighbors.len() == 3 {
                    cell.set_alive();
                   } else if alive_neighbors.len() > 3 {
                    cell.set_dead();
                   }
                } else if cell.is_dead() {
                    if alive_neighbors.len() == 3 {
                        cell.set_alive();
                    }
                }

                // update the cell at the current position on the board
                self.grid[row as usize][col as usize] = cell;
            }
        }
    }

    pub fn get_cell_neighbor_info(&self, cell: Cell) -> (Vec<Cell>, Vec<Cell>) {
        let mut alive_neighbors: Vec<Cell> = Vec::new();
        let mut dead_neighbors: Vec<Cell> = Vec::new();

        let (cell_row, cell_col) = cell.get_coordinates();

        for row in 0..3 {
            for col in 0..3 {
                    let mut neighbor: Cell = cell::get_new_cell();
                    neighbor =  self.grid[cell_row as usize -1 + row][cell_col as usize - 1 + col];
                    if neighbor.get_coordinates() != cell.get_coordinates() {
                        if neighbor.is_alive() {
                            alive_neighbors.push(neighbor);
                        } else {
                            dead_neighbors.push(neighbor);
                        }
                     }
            }
        }
        
        return (alive_neighbors, dead_neighbors);
    }
}

File: PythonVersion/board.py
from random import randint

from cell import Automata


class Board:
    def __init__(self, rows, cols):
        self._rows = rows
        self._cols = cols

        # cell_row = [Automata(alive=False) for i in range(self._rows)]
        # self._grid = [list(cell_row) for j in range(self._cols)]

        self._grid = [
            [
                Automata(x=row_cells, y=column_cells, alive=False)
                for column_cells in range(self._cols)
            ]
            for row_cells in range(self._rows)
        ]

        self.generate_board()

    def __getitem__(self, key):
        x = key[0]
        y = key[1]
        return self._grid[x][y]

    def generate_board(self):
        for row in self._grid:
            for automata in row:
                # there is a 33% chance the cells spawn alive.
                chance_number = randint(0, 2)
                if chance_number == 1:
                    automata.set_alive()

    def show(self):
        print("\n" * 2)
        for row in self._grid:
            for automata in row:
                print(automata, end="")
            print()  # to create a new line pr. row.

    def update(self):
        for row in self._grid:
            for automata in row:
                # Analyze atomatas neighbors
                alive_neighbors, dead_neighbors = self.check_automata_neighbors(
                    automata
                )
                # Alive ruleset
                if automata._alive == True:
                    if len(alive_neighbors) < 2:
                        automata.set_dead()
                    elif len(alive_neighbors) == 2 or len(alive_neighbors) == 3:
                        automata.set_alive()
                    elif len(alive_neighbors) > 3:
                        automata.set_dead()
                # Dead ruleset
                elif automata._alive == False:
                    if len(alive_neighbors) == 3:
                        automata.set_alive()

    def check_automata_neighbors(self, automata):
        alive_neighbors = []
        dead_neighbors = []

        cell_row, cell_col = automata.coordinates()

        for row in range(3):
            for col in range(3):
                try:
                    neighbor = self._grid[cell_row - 1 + row][cell_col - 1 + col]
                    if neighbor.coordinates() != automata.coordinates():
                        if neighbor._alive == True:
                            alive_neighbors.append(neighbor)
                        else:
                            dead_neighbors.append(neighbor)
                except:
                    pass

        return (alive_neighbors, dead_neighbors)
